### ISSUE TITLE
Add more gofailpoints to validate critical parts of the code 

### ISSUE THREAD

**Issue by serathius** [etcd-io/etcd#14735]
### What would you like to be added?

Followup from https://github.com/etcd-io/etcd/pull/14733. Thanks to gofailpoint in defrag code we were able to find an issue with data inconsistency. 

Based on this we should increase coverage of other processes similar to defrag. TODO:
* [ ] Create a list of processes to add gofailpoints
* [ ] Add the gofailpoint into the code (example https://github.com/etcd-io/etcd/blob/c83b1ad9bad637fd02c6e17010755c9b35860bd7/server/storage/backend/backend.go#L504)
* [ ] Add code to linearizability tests to trigger the failpoints

Let's start from creating the list:
* Compact
* AuthEnable
* AuthDisable
* ConsistenIndex

Feel free to suggest other processes that require a gofailpoint. cc @ahrtr @ptabor 

### Why is this needed?

This should allow us to better detect issues with data inconsistency like https://github.com/etcd-io/etcd/pull/14733

**Comment by ahrtr** [comment]
I will provide a list next week.

**Comment by ahrtr** [comment]
Supplement:
- downgrade
- Moveleader
- backend (I will add some failpoints)

**Comment by serathius** [comment]
Would love to also get some suggestions from @tbg @pavelkalinnikov. Raft library already has couple of failpoints, still it might be worth to expand them, especially with https://github.com/etcd-io/etcd/pull/14627.

**Comment by lavacat** [comment]
@serathius I've checked ConsistenIndex and I think the most interesting use case for failpoint is [UnsafeSave](https://github.com/etcd-io/etcd/blob/887f95d0d3818029708773bbe6451e2ebce18a9e/server/etcdserver/cindex/cindex.go#L49), but it's indirectly covered by [PreCommitHook](https://github.com/etcd-io/etcd/blob/6f899a7b40f7631461ffeda0067aa4c42dd17812/server/storage/backend/batch_tx.go#L343) failpoints. 
Do you have something else in mind for ConsistenIndex? If not, I think it can be crossed out.

**Comment by lavacat** [comment]
@ahrtr Checked Moveleader and it seams most of the work is happening inside raft lib. I think it will be useful to add MoveLeader call to [DefaultTraffic](https://github.com/etcd-io/etcd/blob/18a5dc417b84489951de5413b3578462726614be/tests/linearizability/traffic.go#L29). But I don't see a good place inside etcd repo for a new failpoint. Do you have something else in mind for MoveLeader failpoint?

**Comment by serathius** [comment]
> @serathius I've checked ConsistenIndex and I think the most interesting use case for failpoint is [UnsafeSave](https://github.com/etcd-io/etcd/blob/887f95d0d3818029708773bbe6451e2ebce18a9e/server/etcdserver/cindex/cindex.go#L49), but it's indirectly covered by [PreCommitHook](https://github.com/etcd-io/etcd/blob/6f899a7b40f7631461ffeda0067aa4c42dd17812/server/storage/backend/batch_tx.go#L343) failpoints. Do you have something else in mind for ConsistenIndex? If not, I think it can be crossed out.

No, list is to ensure we give some thought to it. Thanks for confirming.

**Comment by serathius** [comment]
I think we should also cover watch with failpoints. First step would be to reproduce https://github.com/etcd-io/etcd/pull/15248 by adding a failpoint after sending a progress notify. cc @ahrtr

**Comment by ahrtr** [comment]
> @ahrtr Checked Moveleader and it seams most of the work is happening inside raft lib. I think it will be useful to add MoveLeader call to [DefaultTraffic](https://github.com/etcd-io/etcd/blob/18a5dc417b84489951de5413b3578462726614be/tests/linearizability/traffic.go#L29). But I don't see a good place inside etcd repo for a new failpoint. Do you have something else in mind for MoveLeader failpoint?

Sorry for the late response. We can use [newLeader](https://github.com/etcd-io/etcd/blob/3c64ae4b7941c5a377e011aae51e5e29da7d0655/server/etcdserver/raft.go#L172) to check whether the leader changed, and add failpoint accordingly.



> I think we should also cover watch with failpoints. First step would be to reproduce #15248 by adding a failpoint after sending a progress notify. cc @ahrtr

This seems to be a good suggestion. But I still do not think it can reliably reproduce the duplicated event issue. Submitted a PR https://github.com/etcd-io/etcd/pull/15272.

**Comment by pranavshenoy** [comment]
@serathius  Getting started with contributing to etd. Shall I take this up?

**Comment by mounilKshah** [comment]
If no one else is working on this issue, is it fine if I work on it @serathius ?

**Comment by serathius** [comment]
This issue requires some research into best places to put failpoints. This might require knowledge from maintainers, that might not be available and not able to provide timely responses.

@mounilKshah  Could you look into https://github.com/etcd-io/etcd/issues/17817 ? It's very similar issue, but about backporting existing failpoints, so there is no need for maintainers help and should allow you to dig into the problem.

**Comment by mounilKshah** [comment]
Yeah sure, thanks for the quick response!
